src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -241,43 +255,60 @@ def frobenius_charpoly(self, var='X', algorithm='crystalline'):
 
         ALGORITHM:
 
-            By default, this method uses the *crystalline* algorithm which
-            computes the characteristic polynomial of the Frobenius acting
-            on the crystalline cohomology of the Drinfeld module. For further
-            details, see [Ang1997]_. Currently, the only alternative is to use
-            the *gekeler* approach based on solving the linear system given by
-            `t^{nr} + \sum_{i=1}^{r} \phi_{A_{i}}t^{n(r-i)} = 0`. For more
-            details, see [Gek2008]_.
-
+        By default, this method uses the so-called *crystalline*
+        algorithm which computes the characteristic polynomial of the
+        Frobenius acting on the crystalline cohomology of the Drinfeld
+        module. For further details, see [Ang1997]_. Currently, the only
+        alternative is to use the *gekeler* approach based on solving
+        the linear system given by `t^{nr} + \sum_{i=1}^{r}
+        \phi_{A_{i}}t^{n(r-i)} = 0`. For more details, see [Gek2008]_.
         """
-        algorithms = {'gekeler', 'crystalline'}
-        if algorithm in algorithms:
-            return getattr(self, \
-                    f'_frobenius_charpoly_{algorithm}')(var)
+        methods = [method for method in dir(self)
+                   if method.startswith('_frobenius_charpoly_')]
+        method_name = f'_frobenius_charpoly_{algorithm}'
+        if method_name in methods:
+            if self._frobenius_charpoly is not None:
+                return self._frobenius_charpoly
+            self._frobenius_charpoly = getattr(self, method_name)(var)
+            return self._frobenius_charpoly
         raise NotImplementedError(f'Algorithm \"{algorithm}\" not implemented')
 
     def _frobenius_charpoly_crystalline(self, var='X'):
```

Comment:
My only comment for this method is the following: I think you did not address my previous comment on the `var='X'` for `_frobenius_charpoly_...` methods. Shouldn't we remove it here? It seems redundant.

The rest of the method is perfect, thank you for the changes!